src/index.tsx

```diff
@@ -27,38 +29,83 @@ import {
-import { MachineContext, MachineProvider } from "./xstate";
-import * as selectors from "./xstate/selectors";
 
 import "maplibre-gl/dist/maplibre-gl.css";
 import "./globals.css";
 
 await initParquetWasm();
 
 function App() {
-  const actorRef = MachineContext.useActorRef();
-  const isDrawingBBoxSelection = MachineContext.useSelector(
-    selectors.isDrawingBBoxSelection,
-  );
-  const isOnMapHoverEventEnabled = MachineContext.useSelector(
-    selectors.isOnMapHoverEventEnabled,
-  );
+  // =========================================================================
+  // Client-Side State (Zustand)
+  // UI-only state that never syncs with Python
+  // See: src/state/store.ts
+  // =========================================================================
 
-  const highlightedFeature = MachineContext.useSelector(
-    (s) => s.context.highlightedFeature,
+  // Feature highlighting
+  const highlightedFeature = useStore((state) => state.highlightedFeature);
+  const setHighlightedFeature = useStore(
+    (state) => state.setHighlightedFeature,
   );
 
-  const bboxSelectPolygonLayer = MachineContext.useSelector(
-    selectors.getBboxSelectPolygonLayer,
-  );
-  const bboxSelectBounds = MachineContext.useSelector(
-    selectors.getBboxSelectBounds,
-  );
+  // Bounding box selection
+  const isDrawingBbox = useStore((state) => state.isDrawingBbox);
+  const bboxSelectStart = useStore((state) => state.bboxSelectStart);
+  const bboxSelectEnd = useStore((state) => state.bboxSelectEnd);
+  const setBboxStart = useStore((state) => state.setBboxStart);
+  const setBboxEnd = useStore((state) => state.setBboxEnd);
+  const setBboxHover = useStore((state) => state.setBboxHover);
+
+  const isOnMapHoverEventEnabled =
+    isDrawingBbox && bboxSelectStart !== undefined;
+
+  const bboxSelectBounds = useMemo(() => {
+    if (bboxSelectStart && bboxSelectEnd) {
+      const [x0, y0] = bboxSelectStart;
+      const [x1, y1] = bboxSelectEnd;
+      return [
+        Math.min(x0, x1),
+        Math.min(y0, y1),
+        Math.max(x0, x1),
+        Math.max(y0, y1),
+      ];
+    }
+    return null;
+  }, [bboxSelectStart, bboxSelectEnd]);
+
+  const bboxSelectPolygonLayer = useMemo(() => {
+    if (bboxSelectStart && bboxSelectEnd) {
+      const bboxProps: PolygonLayerProps = {
+        id: "bbox-select-polygon",
+        data: [
+          [
+            [bboxSelectStart[0], bboxSelectStart[1]],
+            [bboxSelectEnd[0], bboxSelectStart[1]],
+            [bboxSelectEnd[0], bboxSelectEnd[1]],
+            [bboxSelectStart[0], bboxSelectEnd[1]],
+          ],
+        ],
+        getPolygon: (d) => d,
+        getFillColor: [0, 0, 0, 50],
+        getLineColor: [0, 0, 0, 130],
+        stroked: true,
+        getLineWidth: 2,
+        lineWidthUnits: "pixels",
+      };
+      if (isDrawingBbox) {
+        bboxProps.getFillColor = [255, 255, 0, 120];
+        bboxProps.getLineColor = [211, 211, 38, 200];
+        bboxProps.getLineWidth = 2;
+      }
```

Comment:
These can always be set on `bboxProps`. The canonical way to toggle visibility is with the layer's [`visible` prop](https://deck.gl/docs/api-reference/core/layer#visible). So we can set `visible: false` in general and then 
```ts
if (isDrawingBbox) {
  bboxProps.visible = true;
}
```